src/main/java/hudson/plugins/sshslaves/SSHLauncher.java

```diff
@@ -1338,78 +1354,91 @@ private boolean isRecoverable(String message) {
      * {@inheritDoc}
      */
     @Override
-    public synchronized void afterDisconnect(SlaveComputer slaveComputer, final TaskListener listener) {
-        if (connection != null) {
-            boolean connectionLost = reportTransportLoss(connection, listener);
-            if (session!=null) {
-                // give the process 3 seconds to write out its dying message before we cut the loss
-                // and give up on this process. if the slave process had JVM crash, OOME, or any other
-                // critical problem, this will allow us to capture that.
-                // exit code is also an useful info to figure out why the process has died.
-                try {
-                    listener.getLogger().println(getSessionOutcomeMessage(session,connectionLost));
-                    session.getStdout().close();
-                    session.close();
-                } catch (Throwable t) {
-                    t.printStackTrace(listener.error(Messages.SSHLauncher_ErrorWhileClosingConnection()));
+    public void afterDisconnect(SlaveComputer slaveComputer, final TaskListener listener) {
+        if (connection == null) {
+            // Nothing to do here, the connection is not established
+            return;
+        }
+
+        ExecutorService srv = launcherExecutorService;
+        if (srv != null) {
+            // If the service is still running, shut it down and interrupt the operations if any
+            srv.shutdown();
+        }
+
+        synchronized (this) {
+            if (connection != null) {
```

Comment:
Would prefer less reindentation. Even with the `?w=1` diff, there could be issues backporting, merging, blaming, etc.